src/mame/trainer/ms68k.cpp

```diff
@@ -0,0 +1,327 @@
+
+	// Set up printer.
+
+	CENTRONICS(config, m_printer_conn, centronics_devices, "printer");
+	OUTPUT_LATCH(config, m_printer_out);
+	m_printer_conn->set_output_latch(*m_printer_out);
+	// Printer Interrupt = 2
+	m_printer_conn->ack_handler().append_inputline(m_maincpu, M68K_IRQ_2);
+	// DUART IP2 = Busy
+	m_printer_conn->busy_handler().set(m_duart, FUNC(mc68681_device::ip2_w));
+	// DUART IP3 = Paper Out
+	m_printer_conn->perror_handler().set(m_duart, FUNC(mc68681_device::ip3_w));
+	// DUART IP4 = Device Selected
+	m_printer_conn->select_handler().set(m_duart, FUNC(mc68681_device::ip4_w));
+	// DUART IP5 = Acknowledge
+	m_printer_conn->ack_handler().set(m_duart, FUNC(mc68681_device::ip5_w));
+
+	// Set up Expansion.
+	// EXT Interrupt = 6
+	// TODO: Implement expansion bus.
+}
+
+void ms68k_state::mem_map(address_map &map)
+{
+	map.unmap_value_high();
+	map(0x000000, 0x07ffff).ram().share("ram"); // 512KB RAM onboard maximum
+	map(0x000008, 0x07ffff).ram();
+	map(0xd00000, 0xd7ffff).rw(m_scsi, FUNC(ncr5380_device::read), FUNC(ncr5380_device::write)).umask16(0x00ff);
+	map(0xd80000, 0xdfffff).rw(FUNC(ms68k_state::printer_r), FUNC(ms68k_state::printer_w)).umask16(0x00ff);
+	map(0xe00000, 0xe7ffff).rw(m_fdc, FUNC(wd1772_device::read), FUNC(wd1772_device::write)).umask16(0x00ff);
+	map(0xe80000, 0xefffff).r(m_duart, FUNC(mc68681_device::read)).w(FUNC(ms68k_state::duart_w)).umask16(0x00ff);
+	map(0xf00000, 0xf1ffff).rom().region("roms", 0); // 128KB ROM onboard maximum
+
+	// ROM is mapped to 0 on reset, remapped on first DUART write
+	map(0x000000, 0x000007).view(m_bootvect);
+	m_bootvect[0](0x000000, 0x000007).rom().region("roms", 0);
+}
+
+void ms68k_state::machine_start()
+{
+	save_item(NAME(m_floppy_selected_drive));
+}
+
+void ms68k_state::machine_reset()
+{
+	// Reset pointer to bootvector in ROM for bootvector view
+	m_bootvect.select(0);
+}
+
+/* DUART */
+
+void ms68k_state::duart_w(offs_t offset, uint8_t data)
+{
+	m_duart->write(offset, data);
+
+	// The first write to the DUART also swaps ROM and RAM.
+	if (m_bootvect.entry().has_value()) {
+		m_bootvect.disable(); // stop mapping ROM until reset
+	}
+}
+
+/* Floppy */
+
+void ms68k_state::floppy_types(device_slot_interface &device)
+{
+	device.option_add("525sd", FLOPPY_525_SD);
+	device.option_add("525dd", FLOPPY_525_DD);
+	device.option_add("35dd", FLOPPY_35_DD);
+	device.option_add("35hd", FLOPPY_35_HD);
+}
+
+void ms68k_state::floppy_formats(format_registration &fr)
+{
+	floppy_image_device::default_mfm_floppy_formats(fr);
+}
+
+void ms68k_state::floppy_side_w(int state)
+{
+	floppy_image_device *floppy = nullptr;
+	floppy = m_floppy[m_floppy_selected_drive]->get_device();
+	if (floppy) {
+		floppy->ss_w(state);
+	}
+}
+
+void ms68k_state::floppy_drive0_w(int state)
+{
+	int drive = (state == 1) ? 0 : 1;
+	floppy_drive_select(drive);
+}
+
+void ms68k_state::floppy_drive1_w(int state)
+{
+	int drive = (state == 1) ? 1 : 1;
+	floppy_drive_select(drive);
```

Comment:
Is that correct? If yes, just omit that 1st line and do select(1)?

Reply:
Doh! Good catch, thanks!